File: app/src/main/java/com/suzei/minote/utils/executors/AppExecutor.java
package com.suzei.minote.utils.executors;

import android.os.Handler;
import android.os.Looper;

import java.util.concurrent.Executor;
import java.util.concurrent.Executors;

public class AppExecutor {

    private static final int THREAD_COUNT = 3;

    private static AppExecutor sInstance;

    private final Executor mDiskIO;

    private final Executor mMainThread;

    private final Executor mNetworkIO;

    private AppExecutor(Executor mDiskIO, Executor mMainThread, Executor mNetworkIO) {
        this.mDiskIO = mDiskIO;
        this.mMainThread = mMainThread;
        this.mNetworkIO = mNetworkIO;
    }

    public static AppExecutor getInstance() {

        if (sInstance == null) {
            sInstance = new AppExecutor(
                    new DiskIOThreadExecutor(),
                    Executors.newFixedThreadPool(THREAD_COUNT),
                    new MainThreadExecutor());
        }

        return sInstance;
    }

    public Executor getDiskIO() {
        return mDiskIO;
    }

    public Executor getMainThread() {
        return mMainThread;
    }

    public Executor getNetworkIO() {
        return mNetworkIO;
    }

    private static class MainThreadExecutor implements Executor {

        private Handler mainThreadHandler = new Handler(Looper.getMainLooper());

        @Override
        public void execute(Runnable command) {
            mainThreadHandler.post(command);
        }
    }

}

File: app/src/main/java/com/suzei/minote/utils/executors/AppExecutor.kt
package com.suzei.minote.utils.executors

import android.os.Handler
import android.os.Looper
import androidx.annotation.VisibleForTesting
import java.util.concurrent.Executor

class AppExecutor @VisibleForTesting
internal constructor(val diskIO: Executor, val mainThread: Executor) {

    private class MainThreadExecutor : Executor {

        private val mainThreadHandler = Handler(Looper.getMainLooper())

        override fun execute(command: Runnable) {
            mainThreadHandler.post(command)
        }

    }

    companion object {

        val instance: AppExecutor

            get() {

                return AppExecutor(
                        DiskIOThreadExecutor(),
                        MainThreadExecutor())
            }
    }

}
